legend-engine-shared-core/src/main/java/org/finos/legend/engine/shared/core/operational/prometheus/MetricsHandler.java

```diff
@@ -227,42 +221,212 @@ public static synchronized void observeCount(String name, String[] labelNames, S
         }
     }
 
-    @Deprecated
-    public static synchronized void observeErrorCount(String name)
+    // -------------------------------------- ERROR HANDLING -------------------------------------
+
+    /**
+     * Path to JSON file outlining error data for categorisation available to open source.
+     */
+    private static final String ERROR_DATA_PATH = "/ErrorData.json";
+
+    /**
+     * Prometheus counter to record errors with labels of the service causing the error if it is a service-related error,
+     * the label given to the error, the category of the error and source of the error.
+     */
+    protected static final Counter ERROR_COUNTER = Counter.build("legend_engine_error_total", "Count errors in legend ecosystem").labelNames("errorLabel", "category", "source", "serviceName").register(getMetricsRegistry());
+
+    /**
+     * Types of error matching techniques that can be performed on an incoming exceptions.
+     */
+    public enum MatchingMethod
+    { EXCEPTION_OUTLINE_MATCHING, KEYWORDS_MATCHING, TYPE_NAME_MATCHING }
+
+    /**
+     * List of objects corresponding to the error categories holding their associated exception data.
+     */
+    private static final List<ExceptionCategory> ERROR_CATEGORY_DATA_OBJECTS = readErrorData();
+
+    /**
+     * Method to record an error occurring during execution and add it to the metrics.
+     * @param origin the stage in execution at which the error occurred.
+     * @param exception the non-null exception to be analysed that has occurred in execution.
+     * @param servicePath the name of the service whose execution invoked the error.
+     */
+    public static synchronized void observeError(Enum origin, Exception exception, String servicePath)
```

Comment:
Interesting idiom. First time I am seeing this in a code base. So this to allow LoggingEventType to be extended outside legend-engine ?

Reply:
Yes, should work with any enum defined outside of engine - I just added a test for it as well